File: 잡동사니/k2022/spring/연습/meeting/src/main/java/com/karma/meeting/controller/FeedController.java
package com.karma.meeting.controller;

import com.karma.meeting.model.dto.FeedDto;
import com.karma.meeting.model.util.CustomPrincipal;
import com.karma.meeting.service.FeedService;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.web.PageableDefault;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.stereotype.Controller;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.ui.ModelMap;
import org.springframework.web.bind.annotation.*;

@Controller
@Transactional
@RequiredArgsConstructor
@RequestMapping("/feed")
public class FeedController {
    private final FeedService feedService;

    /**
     * 피드 읽기
     * feedPage : 피드 페이지 읽기
     * feedDetail : 피드 1개 읽기
     */
    @GetMapping
    @Transactional(readOnly = true)
    public String FeedPageView(
            @PageableDefault(size = 20, sort = "createdAt", direction = Sort.Direction.DESC) Pageable pageable,
            ModelMap modelMap
    ){
        Page<FeedDto> feeds = feedService.getFeedPage(pageable);
        modelMap.addAttribute("feeds", feeds);
        return "/feed/index";
    }

    @GetMapping("/{feedId}")
    @Transactional(readOnly = true)
    public String FeedView(
            @PathVariable Long feedId,
            ModelMap modelMap
    ){
        FeedDto feedDto = feedService.getFeed(feedId);
        modelMap.addAttribute("feed", feedDto);
        return "/feed/detail/index";
    }

    /**
     * 피드 추가
     * prefix : add
     * addFeedPage : 피드 쓰기 페이지
     * addFeed : 피드 쓰기 요청처리
     */
    @GetMapping("/add")
    public String addFeedView(){
        return "/feed/add/index";
    }

    @PostMapping("/add")
    public String addFeed(
            @RequestParam("title") String title,
            @RequestParam("content") String content,
            @AuthenticationPrincipal CustomPrincipal principal
    ){
        feedService.addFeed(title, content, principal.getUsername());
        return "redirect:/feed";
    }

    /**
     * 피드 수정
     * prefix : modify
     */

    @GetMapping("/modify/{feedId}")
    public String modifyFeedView(
            @PathVariable Long feedId,
            ModelMap modelMap
    ){
        FeedDto dto = feedService.getFeed(feedId);
        modelMap.addAttribute("feedId", feedId);
        modelMap.addAttribute("title", dto.getTitle());
        return "/feed/modify/index";
    }

    @PostMapping("/modify/{feedId}")
    public String modifyFeed(
            @PathVariable Long feedId,
            @RequestParam("title") String title,
            @RequestParam("content") String content
    ){
        feedService.modifyFeed(feedId, title, content);
        return "redirect:/feed";
    }

    /**
     * 피드 삭제
     * prefix : delete
     */
    @PostMapping("/delete/{feedId}")
    public String deleteFeed(
            @PathVariable Long feedId
    ){
        feedService.deleteFeed(feedId);
        return "redirect:/feed";
    }
}

File: backend/src/main/kotlin/com/karma/chat/controller/FeedController.kt
package com.karma.chat.controller


import com.karma.chat.controller.dto.BodyDto
import com.karma.chat.controller.dto.feed.CreateFeedRequestDto
import com.karma.chat.controller.dto.feed.ModifyFeedRequestDto
import com.karma.chat.domain.feed.Feed
import com.karma.chat.service.FeedService
import com.karma.chat.util.UserAuthorize
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PageableDefault

import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.security.core.userdetails.User
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/feed")
@UserAuthorize
class FeedController(
    private val feedService: FeedService
) {

    @GetMapping
    fun fetchFeeds(
        @PageableDefault pageable: Pageable
    ): ResponseEntity<BodyDto<Iterable<Feed>>> {
        try {
            val feeds = feedService.fetchFeeds(pageable)
            return ResponseEntity.ok(BodyDto(message = "success", data = feeds))
        } catch (e: Exception) {
            e.printStackTrace();
            return ResponseEntity.status(500).body(BodyDto(message = "fetch feed fail"))
        }
    }

    @PostMapping
    fun createFeed(
        @RequestBody req: CreateFeedRequestDto
    ): ResponseEntity<BodyDto<Long>> {
        try {
            val feedId = feedService.createFeed(content = req.content, media = req.media)
            return ResponseEntity.ok(BodyDto(message = "success", data = feedId))
        } catch (e: Exception) {
            e.printStackTrace();
            return ResponseEntity.status(500).body(BodyDto(message = "create feed fail"))
        }
    }

    @PutMapping
    fun modifyFeed(
        @RequestBody req: ModifyFeedRequestDto,
        @AuthenticationPrincipal user: User
    ): ResponseEntity<BodyDto<Long>> {
        try {
            val feedId = feedService.modifyFeed(
                feedId = req.feedId,
                content = req.content,
                media = req.media,
                username = user.username
            )
            return ResponseEntity.ok(BodyDto(message = "success", data = feedId))
        } catch (e: Exception) {
            e.printStackTrace();
            return ResponseEntity.status(500).body(BodyDto(message = "delete feed fail"))
        }
    }

    @DeleteMapping("/{feedId}")
    fun deleteFeed(
        @PathVariable feedId: Long,
        @AuthenticationPrincipal user: User
    ): ResponseEntity<BodyDto<Long>> {
        try {
            feedService.deleteFeed(feedId = feedId, username = user.username)
            return ResponseEntity.ok(BodyDto(message = "success", data = feedId))
        } catch (e: Exception) {
            e.printStackTrace();
            return ResponseEntity.status(500).body(BodyDto(message = "delete feed fail"))
        }
    }
}